### DISCUSSION TITLE
v0.9.0-beta.0

### DISCUSSION THREAD

**Discussion by gadomski** [gadomski/las-rs#94] [Announcements]
Hey folks, I've done some significant refactoring of the library in preparation for a v0.9.0 release. Because there's so many changes, I thought I'd do a beta first to get feedback. A listing of the changes is in the [CHANGELOG](https://github.com/gadomski/las-rs/blob/main/CHANGELOG.md) and is copied below. Your feedback would be appreciated — I've done my best to deprecate instead of remove, but I'm sure something will break. 🍻 

I'll probably wait a week or so before releasing v0.9.0 (if there's no significant changes to the release).

### Added

- Several methods in `Reader` ([#89](https://github.com/gadomski/las-rs/pull/89))
  - `Reader::read_point`, which returns a `Result<Option<Point>>`
  - `Reader::read_points`
  - `Reader::read_points_into`
  - `Reader::read_all_points_into`
- `Writer::write_point`, `Header::write_to`, `laz` module, a few laz-specific methods on `Header` ([#90](https://github.com/gadomski/las-rs/pull/90))

### Changed

- Reorganize reading, including removing the lifetime specifier on `Reader` ([#89](https://github.com/gadomski/las-rs/pull/89))
- Conslidate errors to a single enum ([#87](https://github.com/gadomski/las-rs/pull/87))

### Fixed

- Start of first EVLR ([#91](https://github.com/gadomski/las-rs/pull/91))

## Deprecated

- `Read` trait ([#88](https://github.com/gadomski/las-rs/pull/88))
- `Write` trait ([#90](https://github.com/gadomski/las-rs/pull/90))
- Many methods on `Reader` ([#89](https://github.com/gadomski/las-rs/pull/89))
  - `read` in favor of `read_point`
  - `read_n` in favor of `read_points`
  - `read_n_into` in favor of `read_points_into`
  - `read_all_points` in favor of `read_all_points_into`
- `Writer::write` ([#90](https://github.com/gadomski/las-rs/pull/90))

**Full Changelog**: https://github.com/gadomski/las-rs/compare/v0.8.8...v0.9.0-beta.0

<hr /><em>This discussion was created from the release <a href='https://github.com/gadomski/las-rs/releases/tag/v0.9.0-beta.0'>v0.9.0-beta.0</a>.</em>

**Comment by jrjdavidson** [comment]
Kia ora Pete,

Just tried to upgrade this to 0.9.1 from 0.8.8 in a small script I had written. When trying to pass a reader to closure (i'm trying to use a threads) I get the following error:

the trait `Send` is not implemented for `(dyn reader::ReadPoints + 'static)`, which is required by `{closure@src\lib.rs:164:24: 164:54}: Send`

It's probably a mistake on my end but just leaving this here in case you have some ideas!

Cheers

**Reply by jrjdavidson** [reply]
I updated my script to create the reader in the thread ( by passing the las path to the thread) and it works fine now.
